File: 1209-remove-all-adjacent-duplicates-in-string-ii/1209-remove-all-adjacent-duplicates-in-string-ii.cpp
class Solution {
public:
    string removeDuplicates(string s, int k) {
        stack<pair<char, int>> stk;
        stk.push({'#', 0});
        for (int i = 0; i < size(s); i++) {
            stk.push({s[i], stk.top().first == s[i] ? stk.top().second + 1 : 1});
            if (stk.top().second == k) {
                while(stk.top().first == s[i]) stk.pop();
                s.erase(i - k + 1, k);
                i -= k;
            }
        }
        return s;
    }
};

File: 1209-remove-all-adjacent-duplicates-in-string-ii/1209-remove-all-adjacent-duplicates-in-string-ii.java
class Solution {
    public String removeDuplicates(String s, int k) {
        Stack<Pair<Character, Integer>> stack = new Stack();
        for (char c : s.toCharArray()) {
            if (stack.isEmpty() || c != stack.peek().getKey())
                stack.push(new Pair(c, 1));
            else {
                int cnt = stack.pop().getValue() + 1;
                if (cnt != k)
                    stack.push(new Pair(c, cnt));
            }
        }
        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            int cnt = stack.peek().getValue();
            char c = stack.pop().getKey();
            while (cnt-- > 0)
                sb.append(c);
        }
        return sb.reverse().toString();
    }
}